pkg/metrics/metrics.jsx

```diff
@@ -440,17 +474,68 @@ class CurrentMetrics extends React.Component {
         }
 
         // top 5 CPU and memory consuming systemd units
-        newState.topServicesCPU = n_biggest(this.cgroupCPUNames, this.samples[9], 5).map(
-            ([key, value, is_user, is_container]) => cgroupRow(key, Number(value / 10 / numCpu).toFixed(1), is_user, is_container) // usec/s → percent
+        const topServicesCPU = n_biggest(this.cgroupCPUNames, this.samples[9], 5);
+        newState.topServicesCPU = topServicesCPU.map(
+            ([key, value, is_user, is_container, userid]) => cgroupRow(key, Number(value / 10 / numCpu).toFixed(1), is_user, is_container, userid, this.state.podNameMapping) // usec/s → percent
         );
 
-        newState.topServicesMemory = n_biggest(this.cgroupMemoryNames, this.samples[10], 5).map(
-            ([key, value, is_user, is_container]) => cgroupRow(key, cockpit.format_bytes(value), is_user, is_container)
+        const topServicesMemory = n_biggest(this.cgroupMemoryNames, this.samples[10], 5);
+        newState.topServicesMemory = topServicesMemory.map(
+            ([key, value, is_user, is_container, userid]) => cgroupRow(key, cockpit.format_bytes(value), is_user, is_container, userid, this.state.podNameMapping)
         );
 
+        const notMappedContainers = topServicesMemory.concat(topServicesCPU).filter(([key, value, is_user, is_container, userid]) => is_container && !podNameInCache(this.state.podNameMapping, userid, key));
+        if (notMappedContainers.length !== 0) {
+            this.update_podman_name_mapping(notMappedContainers);
+        }
```

Comment:
You need to take negative cache lookups into account here. I.e. if you previously failed to map an ID to a pod name (perhaps it's a docker/LXC/etc. container or custom cgroup), the cache must have an entry for that ID, mapping it to `null` (whereas "not attempted yet" would be `undefined` or even "key does not exist", which is by and large `undefined` as well). This should then *not* cause another cache refresh.

That should work fine wiht the cache building below, as that only ever updates podNameMapping instead of rewriting it.